### ISSUE TITLE
Type checker for class

### ISSUE THREAD

**Issue by ikatyang** [Microsoft/TypeScript#15547]
**TypeScript Version:**  2.3.2 / nightly (2.4.0-dev.20170502)

**Code**
```ts
// test.ts

class A {}
class B {}

function test(a: A) {
    if (a instanceof A) {
        // typeof a === A
        console.log('always');
    } else {
        // typeof a === never
        console.log('never');
    }
}

const b = new B();
test(b); // should show error message since b is not an instance of A
```

```sh
$ node_modules/.bin/tsc test.ts
$ node test.js
never
```

It seems TS only checks its shape without its prototype chain?

**Expected behavior:**
show an error message

**Actual behavior:**
non-error

**Comment by fatcerberus** [comment]
Under structural typing, `A` and `B` are equivalent.  TypeScript doesn't have nominal types.

**Comment by ikatyang** [comment]
Thanks for the explaining.

It seems I have to add some unique property for TS to check its prototype chain.

```ts
class A { protected _this_is_an_instance_of_A_: any }
class B { protected _this_is_an_instance_of_B_: any }
```

If so, I'd like to suggest this step to be automatic.

**Comment by RyanCavanaugh** [comment]
This is addressed in several different sections of the FAQ
